Speed up fx graph iteration by implementing it in C++

oulgen opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #128288

Before this change
```
python benchmarks/dynamo/microbenchmarks/fx_microbenchmarks.py
iterating over 100000000 FX nodes took 19.5s (5132266 nodes/s)
```

After this change
```
python benchmarks/dynamo/microbenchmarks/fx_microbenchmarks.py
iterating over 100000000 FX nodes took 3.4s (29114001 nodes/s)
```

5.7x improvement

Differential Revision: [D58343997](https://our.internmc.facebook.com/intern/diff/D58343997)

oulgen committed:
Speed up fx graph iteration by implementing it in C++

[ghstack-poisoned]

XuehaiPan commented on torch/csrc/fx/node.cpp:
```diff
@@ -0,0 +1,225 @@
+
+static int NodeBase_init_fn(NodeBase* self, PyObject* args, PyObject* kwds) {
+  self->_erased = false;
+  self->_prev = self;
+  self->_next = self;
```
Should take care of the reference count.

XuehaiPan commented on torch/_C/__init__.pyi.in:
```diff
@@ -2331,3 +2331,14 @@ def _save_pickle(obj: Any) -> bytes: ...
 def _jit_to_static_module(graph_or_module: Union[Graph,ScriptModule]) -> Any: ...
 def _fuse_to_static_module(graph_or_module: Union[Graph,ScriptModule], min_size: _int) -> Any: ...
+
+# Defined in torch/csrc/fx/node.cpp
+class NodeBase:
+    _erased: bool
+    _prev: NodeBase
+    _next: NodeBase
```
```suggestion
# Defined in torch/csrc/fx/node.cpp
class NodeBase:
    _erased: bool
    _prev: "NodeBase"
    _next: "NodeBase"
```

oulgen committed:
Update on "Speed up fx graph iteration by implementing it in C++"


Before this change
```
python benchmarks/dynamo/microbenchmarks/fx_microbenchmarks.py
iterating over 100000000 FX nodes took 19.5s (5132266 nodes/s)
```

After this change
```
python benchmarks/dynamo/microbenchmarks/fx_microbenchmarks.py
iterating over 100000000 FX nodes took 3.4s (29114001 nodes/s)
```

5.7x improvement


[ghstack-poisoned]

oulgen committed:
Update on "Speed up fx graph iteration by implementing it in C++"


Before this change
```
python benchmarks/dynamo/microbenchmarks/fx_microbenchmarks.py
iterating over 100000000 FX nodes took 19.5s (5132266 nodes/s)
```

After this change
```
python benchmarks/dynamo/microbenchmarks/fx_microbenchmarks.py
iterating over 100000000 FX nodes took 3.4s (29114001 nodes/s)
```

5.7x improvement


[ghstack-poisoned]

oulgen committed:
Update on "Speed up fx graph iteration by implementing it in C++"


Before this change
```
python benchmarks/dynamo/microbenchmarks/fx_microbenchmarks.py
iterating over 100000000 FX nodes took 19.5s (5132266 nodes/s)
```

After this change
```
python benchmarks/dynamo/microbenchmarks/fx_microbenchmarks.py
iterating over 100000000 FX nodes took 3.4s (29114001 nodes/s)
```

5.7x improvement


[ghstack-poisoned]

oulgen committed:
Update on "Speed up fx graph iteration by implementing it in C++"


Before this change
```
python benchmarks/dynamo/microbenchmarks/fx_microbenchmarks.py
iterating over 100000000 FX nodes took 19.5s (5132266 nodes/s)
```

After this change
```
python benchmarks/dynamo/microbenchmarks/fx_microbenchmarks.py
iterating over 100000000 FX nodes took 3.4s (29114001 nodes/s)
```

5.7x improvement


[ghstack-poisoned]

oulgen commented on torch/csrc/fx/node.cpp:
```diff
@@ -0,0 +1,241 @@
+  return self;
+}
+
+static int NodeBase_init_fn(NodeBase* self, PyObject* args, PyObject* kwds) {
+  self->_erased = false;
+  Py_INCREF(self);
+  self->_prev = self;
+  Py_INCREF(self);
+  self->_next = self;
+  return 0;
+}
+
+static void NodeBase_dealloc(NodeBase* self) {
+  Py_XDECREF(self->_prev);
+  Py_XDECREF(self->_next);
+  Py_TYPE(self)->tp_free((PyObject*)self);
+}
```
There appears to be a memory leak, I have done some ablations in order to figure out the cause, it shows that it is coming from NodeBase, and most specifically from `_prev` and `_next`. I suspect them being initialiazed to self is not jiving well with the gc somehow but not really sure how to prove. If anyone has an idea on how to figure this out, i'd appreciate it.

ezyang replied:
Well, I mean, if you initialize the field to self then you have a reference cycle, and it won't get eliminated until you GC, which could trigger a weakref test

oulgen replied:
This was my first time writing something with c api, did you mean someone how marking these pointers as weak ref? any docs on this? Googling "python c api weak ref" did not give me much

ezyang replied:
Oh but I guess your point is the old code also generated cycles and was not a problem

ezyang replied:
Check https://github.com/pytorch/pytorch/pull/128288/files#r1632406879

XuehaiPan replied:
You could refer to the implementation of `_functools.lru_list_item` and `_functools.lru_cache_object` for reference.

https://github.com/python/cpython/blob/35c799d79177b962ddace2fa068101465570a29a/Modules/_functoolsmodule.c#L761-L776

ezyang commented on torch/fx/graph.py:
```diff
@@ -4,6 +4,7 @@
 import torch.utils._pytree as pytree
 from . import _pytree as fx_pytree
 from ._compatibility import compatibility
+from torch._C import _NodeIter
```
Do you think we ought to preserve the old Python implementation, similar to how some CPython modules have both pure Python and C versions?

oulgen replied:
When would we use the pure python versions? I can implement classes that do exactly the same in python but I assume we will keep moving parts of fx to c++

ezyang replied:
I assume it's more as a reference implementation / spec / testing utility. IDK, I don't feel strongly about it.

ezyang commented on torch/fx/node.py:
```diff
@@ -247,6 +246,16 @@ def __init__(self, graph: 'Graph', name: str, op: str, target: 'Target',
+        self.__dict__.update(d)
+        self._erased = erased
+        self._prev = prev
+        self._next = next
```
Is this BC with the old pickle format?

oulgen replied:
Yes, it would be possible to implement the pickling for NodeBase in C++ and call that here but i didnt think it was necessary

ezyang replied:
no, I don't care about having this in C++, just that it's BC/FC

jansel replied:
I don't believe this is BC with old format.  Since the old format was not a tuple of 4 items.

oulgen replied:
Should I shove erased, prev, next into the `__dict__` and then remove them in setstate?

ezyang commented on torch/_C/__init__.pyi.in:
```diff
@@ -2332,3 +2332,14 @@ def _save_pickle(obj: Any) -> bytes: ...
+
+# Defined in torch/csrc/fx/node.cpp
+class _NodeBase:
+    _erased
```
type annot?

ezyang commented on torch/csrc/fx/node.cpp:
```diff
@@ -0,0 +1,241 @@
+///////////////////////////////
+
+struct NodeBase {
+  PyObject_HEAD bool _erased;
```
clang format disaster

ezyang commented on torch/csrc/fx/node.cpp:
```diff
@@ -0,0 +1,241 @@
+    nullptr, /* tp_as_buffer */
+    Py_TPFLAGS_DEFAULT, /* tp_flags */
+    nullptr, /* tp_doc */
+    nullptr, /* tp_traverse */
```
You haven't implemented tp_traverse for this type. Therefore, the GC cannot run on it and break cycles

oulgen replied:
Nice catch! let me try this

ezyang commented:
I am lazy and hoping to foist the line by line check every refcount is done correctly review on someone else LOL

oulgen committed:
Update on "Speed up fx graph iteration by implementing it in C++"


Before this change
```
python benchmarks/dynamo/microbenchmarks/fx_microbenchmarks.py
iterating over 100000000 FX nodes took 19.5s (5132266 nodes/s)
```

After this change
```
python benchmarks/dynamo/microbenchmarks/fx_microbenchmarks.py
iterating over 100000000 FX nodes took 3.4s (29114001 nodes/s)
```

5.7x improvement


[ghstack-poisoned]

oulgen commented:
@oulgen has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D58343997).

jansel commented on torch/fx/node.py:
```diff
@@ -197,6 +198,7 @@ def __init__(self, graph: 'Graph', name: str, op: str, target: 'Target',
                 annotation of values in the generated code or for other types
                 of analyses.
         """
+        super().__init__()
```
Not passing any args to super?

oulgen replied:
Init just sets erased, prev and next. I dont think I need to pass anything else.

oulgen committed:
Update on "Speed up fx graph iteration by implementing it in C++"


Before this change
```
python benchmarks/dynamo/microbenchmarks/fx_microbenchmarks.py
iterating over 100000000 FX nodes took 19.5s (5132266 nodes/s)
```

After this change
```
python benchmarks/dynamo/microbenchmarks/fx_microbenchmarks.py
iterating over 100000000 FX nodes took 3.4s (29114001 nodes/s)
```

5.7x improvement

Differential Revision: [D58343997](https://our.internmc.facebook.com/intern/diff/D58343997)

[ghstack-poisoned]

oulgen committed:
Update on "Speed up fx graph iteration by implementing it in C++"


Before this change
```
python benchmarks/dynamo/microbenchmarks/fx_microbenchmarks.py
iterating over 100000000 FX nodes took 19.5s (5132266 nodes/s)
```

After this change
```
python benchmarks/dynamo/microbenchmarks/fx_microbenchmarks.py
iterating over 100000000 FX nodes took 3.4s (29114001 nodes/s)
```

5.7x improvement

Differential Revision: [D58343997](https://our.internmc.facebook.com/intern/diff/D58343997)

[ghstack-poisoned]

albanD commented on torch/csrc/fx/node.cpp:
```diff
@@ -0,0 +1,269 @@
+  NodeBase* _next;
+};
+
+static PyObject* NodeBase_new(
```
Isn't that the default implementation? I would expect you don't need to provide it if you don't have special handling.

oulgen replied:
If I remove this implementation, I get
```
TypeError: cannot create 'Node' instances
```

albanD replied:
Ho right they don't inherit it. Ok!

ezyang commented on torch/csrc/fx/node.cpp:
```diff
@@ -0,0 +1,269 @@
+  Py_TYPE(self)->tp_free(self);
+}
+
+static PyTypeObject NodeIterType = {
```
We're not doing the binding in pybind11 because speed is paramount

albanD replied:
But my guess is that the main speedup here is to do the iterations via direct inline access instead of a full python opcode every time. Is the actual cost of calling from python->c++ that big here?
I am not suggesting we should rewrite this PR that already exists! I was just curious if we tried it and if so what were the numbers.

ezyang replied:
It just seems weird to me to say, "Hey, for this microoptimization, have you tried benchmarking it with X, where X is something we *know* is really slow." Like hey, if your goal is to speed things up, don't do things you know are slow!!

oulgen replied:
Yep, doing this via PyBind would completely defeat this PR

albanD replied:
But if you get 95% of the speedup with 1/3 of the code, it takes you half the time to land and it is less likely to contain bugs that sounds like a valid tradeoff to consider? Am I missing something here?

oulgen replied:
I'm very doubtful that you'll get 95% of the speedup. The overhead of calling into pybind will likely even turn this into a regression. @jansel ?

Since I already implemented this, I don't really want to implement it again in pybind just so that we can prove that it is much worse in performance :/

albanD replied:
> Since I already implemented this, I don't really want to implement it again in pybind just so that we can prove that it is much worse in performance :/

Of course, as I said above, this was really not what I was asking for here.
I was asking in case you had that data so I could improve my mental model of how, when and by how much is pybind slower than manual bindings.

oulgen commented on torch/csrc/fx/node.cpp:
```diff
@@ -0,0 +1,269 @@
+
+PyObject* NodeIter_iternext(PyObject* _self) {
+  NodeIter* self = (NodeIter*)_self;
+  if (self->_reversed) {
+    return NodeIter_iternext_helper<true>(self);
+  } else {
+    return NodeIter_iternext_helper<false>(self);
+  }
```
Pretty sure the value of template needs to be known at compile time

albanD replied:
Right, so the compiler is not optimistic enough to generate the two variants valid for a bool? That's fair.

oulgen committed:
Update on "Speed up fx graph iteration by implementing it in C++"


Before this change
```
python benchmarks/dynamo/microbenchmarks/fx_microbenchmarks.py
iterating over 100000000 FX nodes took 19.5s (5132266 nodes/s)
```

After this change
```
python benchmarks/dynamo/microbenchmarks/fx_microbenchmarks.py
iterating over 100000000 FX nodes took 3.4s (29114001 nodes/s)
```

5.7x improvement

Differential Revision: [D58343997](https://our.internmc.facebook.com/intern/diff/D58343997)

[ghstack-poisoned]

oulgen committed:
Update on "Speed up fx graph iteration by implementing it in C++"


Before this change
```
python benchmarks/dynamo/microbenchmarks/fx_microbenchmarks.py
iterating over 100000000 FX nodes took 19.5s (5132266 nodes/s)
```

After this change
```
python benchmarks/dynamo/microbenchmarks/fx_microbenchmarks.py
iterating over 100000000 FX nodes took 3.4s (29114001 nodes/s)
```

5.7x improvement

Differential Revision: [D58343997](https://our.internmc.facebook.com/intern/diff/D58343997)

[ghstack-poisoned]

oulgen commented:
@albanD found https://github.com/python/cpython/blob/main/Objects/iterobject.c#L171 which uses `PyObject_SelfIter`

aorenste commented on torch/_C/__init__.pyi.in:
```diff
@@ -2332,3 +2332,14 @@ def _save_pickle(obj: Any) -> bytes: ...
+
+class _NodeIter:
+    def __init__(self, root: _NodeBase, reversed: _bool) -> None: ...
+    def __iter__(self): ...
```
```
def __iter__(self) -> Iterable[_NodeBase]: ...
```

aorenste commented on torch/_C/__init__.pyi.in:
```diff
@@ -2332,3 +2332,14 @@ def _save_pickle(obj: Any) -> bytes: ...
+    _prev: "_NodeBase"
+    _next: "_NodeBase"
+
+class _NodeIter:
```
```
class _NodeIter(Iterator):
```

aorenste commented on torch/csrc/fx/node.cpp:
```diff
@@ -0,0 +1,256 @@
+    }
+    if constexpr (reversed) {
+      NodeBase* prev = (NodeBase*)Py_NewRef(self->_cur->_prev);
+      Py_XDECREF(self->_cur);
```
consider Py_CLEAR? I think this is reachable via GC.  Or consider:
```
std::swap(self->_cur, prev);
Py_XDECREF(prev);
```
here and below

oulgen committed:
Update on "Speed up fx graph iteration by implementing it in C++"


Before this change
```
python benchmarks/dynamo/microbenchmarks/fx_microbenchmarks.py
iterating over 100000000 FX nodes took 19.5s (5132266 nodes/s)
```

After this change
```
python benchmarks/dynamo/microbenchmarks/fx_microbenchmarks.py
iterating over 100000000 FX nodes took 3.4s (29114001 nodes/s)
```

5.7x improvement

Differential Revision: [D58343997](https://our.internmc.facebook.com/intern/diff/D58343997)

[ghstack-poisoned]

oulgen committed:
Update on "Speed up fx graph iteration by implementing it in C++"


Before this change
```
python benchmarks/dynamo/microbenchmarks/fx_microbenchmarks.py
iterating over 100000000 FX nodes took 19.5s (5132266 nodes/s)
```

After this change
```
python benchmarks/dynamo/microbenchmarks/fx_microbenchmarks.py
iterating over 100000000 FX nodes took 3.4s (29114001 nodes/s)
```

5.7x improvement

Differential Revision: [D58343997](https://our.internmc.facebook.com/intern/diff/D58343997)

[ghstack-poisoned]

oulgen committed:
Update on "Speed up fx graph iteration by implementing it in C++"


Before this change
```
python benchmarks/dynamo/microbenchmarks/fx_microbenchmarks.py
iterating over 100000000 FX nodes took 19.5s (5132266 nodes/s)
```

After this change
```
python benchmarks/dynamo/microbenchmarks/fx_microbenchmarks.py
iterating over 100000000 FX nodes took 3.4s (29114001 nodes/s)
```

5.7x improvement

Differential Revision: [D58343997](https://our.internmc.facebook.com/intern/diff/D58343997)

[ghstack-poisoned]

oulgen committed:
Update on "Speed up fx graph iteration by implementing it in C++"


Before this change
```
python benchmarks/dynamo/microbenchmarks/fx_microbenchmarks.py
iterating over 100000000 FX nodes took 19.5s (5132266 nodes/s)
```

After this change
```
python benchmarks/dynamo/microbenchmarks/fx_microbenchmarks.py
iterating over 100000000 FX nodes took 3.4s (29114001 nodes/s)
```

5.7x improvement

Differential Revision: [D58343997](https://our.internmc.facebook.com/intern/diff/D58343997)

[ghstack-poisoned]

albanD commented on torch/csrc/fx/node.cpp:
```diff
@@ -0,0 +1,269 @@
+
+PyObject* NodeIter_iter(PyObject* self) {
+  Py_INCREF(self);
+  return self;
```
From a bit more searching, this should mimic https://docs.python.org/3/c-api/object.html#c.PyObject_GetIter which seems to be fine with returning itself if self is already an iterator.

To answer my question above, iterating over this object twice is not expected as this object is created as the iterator).
I think the "classical" thing to do here would be that you already have a holder (like a list of nodes) on which you implement tp_iter and that will return the iter object.
But you don't have that and you don't want to directly iterator over the nodes (I guess at least because you want to be able to specify the reversed argument). So you manually create this iterator that is only ever used once and only for iterating.

oulgen commented:
@albanD @ezyang @jansel I think all comments are addressed (I don't think pybind implementation will be anywhere near as performant so I'd prefer not implementing it again for comparison). Can we continue with this PR or is there something y'all would like to see?

albanD reviewed (approved):
Sounds good! Thanks for the updates!

oulgen committed:
Update on "Speed up fx graph iteration by implementing it in C++"


Before this change
```
python benchmarks/dynamo/microbenchmarks/fx_microbenchmarks.py
iterating over 100000000 FX nodes took 19.5s (5132266 nodes/s)
```

After this change
```
python benchmarks/dynamo/microbenchmarks/fx_microbenchmarks.py
iterating over 100000000 FX nodes took 3.4s (29114001 nodes/s)
```

5.7x improvement

Differential Revision: [D58343997](https://our.internmc.facebook.com/intern/diff/D58343997)

[ghstack-poisoned]

oulgen commented:
@pytorchbot merge
